agreement/player.go

```diff
@@ -262,23 +266,39 @@ func (p *player) handleThresholdEvent(r routerHandle, e thresholdEvent) []action
 	var actions []action
 	if e.t() == certThreshold {
 		// this threshold must be for p.Round, and originates from the vote SM tree
-		cert := Certificate(e.Bundle)
+		// First, stage the threshold if not already. It is entirely fine to stage values for periods in the past.
+		r.dispatch(*p, e, proposalMachine, p.Round, p.Period, 0)
+		// Now, also check if we have the block.
 		res := stagedValue(*p, r, e.Round, e.Period)
```

Comment:
After discussion, should be p.Period so that we're reading the staged value from the current period

Reply:
Actually, `e.Round` and `e.Period` are correct; the router sends the thresholds based on `e.Round` and `e.Period` to the right place, and `p.Round` and `p.Period` are misleading...however, the system must still ensure that proposals in old periods are not dropped.